ProductButton: Add animation

gigitux opened the pull request:
This PR implements the animation discussed into https://github.com/woocommerce/woocommerce-blocks/discussions/10153#discussioncomment-6475505.

| Cached Page | Not Cached Page |
|--------|--------|
|<video src=https://github.com/woocommerce/woocommerce-blocks/assets/4463174/3da139cf-7cd6-4b14-8e80-41af607b0e39 />|<video src=https://github.com/woocommerce/woocommerce-blocks/assets/4463174/f8b16c63-49a7-4448-8553-3c81b8d85844 />| 




### Testing

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Install the plugin [WP Optimize](https://wordpress.org/plugins/wp-optimize/).
2. Ensure that your cart is empty.
3. Ensure that you have the Mini Cart block in the header.
4. Enable the cache.
5. Visit the page with the Products block.
6. Refresh the page.
7. Add a product.
8. Refresh the page.
9. Ensure that the counter is updated with a smoother animation.
10. Open the Mini Cart block and change the quantity of the product in the cart.
11. Ensure that the counter inside the button is updated with a smoother animation.
12. Disable the plugin.
13. Visit the page with the Products block.
14. Open the Mini Cart block and change the quantity of the product in the cart.
15. Ensure that the counter inside the button is updated with a smoother animation.
16. Ensure that there isn't any regression.

gigitux committed:
implement animation

gigitux commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -19,45 +19,81 @@ type Context = {
+				return ref.textContent;
+			}
+
+			ref.textContent = '';
```
This is necessary otherwise, the returned string (line 97) is appended.

Aljullu commented on assets/js/atomic/blocks/product-elements/button/style.scss:
```diff
@@ -50,6 +61,17 @@
 		word-break: break-word;
+		width: 150px;
+
+		span {
+
+			&.wc-block-scrollToUp {
+				animation: scrollToUp 0.1s linear 1 normal;
+			}
+			&.wc-block-scrollFromDown {
+				animation: scrollFromDown 0.1s linear 1 normal;
+			}
+		}
```
This should probably be wrapped with [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion).

Aljullu commented on src/BlockTypes/ProductButton.php:
```diff
@@ -115,7 +115,7 @@ protected function render( $attributes, $content, $block ) {
 				'woocommerce' => array(
 					'isLoading'     => false,
-					'numberOfItems' => $number_of_items_in_cart,
+					'numberOfItems' => strval( $number_of_items_in_cart ),
```
I think there is an inconsistency between this line and L157. There we set this property to be an int:

```
$context['woocommerce']['numberOfItems'] = 0;
```

Aljullu commented on src/BlockTypes/ProductButton.php:
```diff
@@ -144,53 +144,50 @@ protected function render( $attributes, $content, $block ) {
-				),
+			$span_button_directives = 'data-wc-text="selectors.woocommerce.addToCartText"';
+
+			if ( WC()->cart->is_empty() ) {
+				$context['woocommerce']['numberOfItems'] = 0;
+				$div_directives                          = 'data-wc-context=\'' . wp_json_encode( $context, JSON_NUMERIC_CHECK ) . '\'';
```
Nit: could this be rewritten to something like:

```PHP
if ( WC()->cart->is_empty() ) {
	$context['woocommerce']['numberOfItems'] = 0;
	$div_directives = 'data-wc-context=\'' . wp_json_encode( $context, JSON_NUMERIC_CHECK ) . '\'';
} else {
	$this->prevent_cache();
}

return apply_filters(
	'woocommerce_loop_add_to_cart_link',
	...
);
```

I didn't test it, but this way you wouldn't need to create the `get_button_template()` method.

Aljullu commented on assets/js/atomic/blocks/product-elements/button/style.scss:
```diff
@@ -42,6 +33,26 @@
 		display: none;
 	}
 
+	@keyframes scrollToUp {
+		from {
+			transform: translateY(0);
+		}
+		to {
+			transform: translateY(-60%);
+		}
+	}
+
+	@keyframes scrollFromDown {
```
I feel a bit confusing to use the word `scroll` for these animations, I usually relate "scroll" to scrolling the page, not to elements animating up or down. I would maybe name them "slideOut" and "slideIn", which also make it clear which element is visible at the beginning and which one at the end of the animation. What do you think?

Aljullu commented on src/BlockTypes/ProductButton.php:
```diff
@@ -144,53 +144,50 @@ protected function render( $attributes, $content, $block ) {
-				),
+			$span_button_directives = 'data-wc-text="selectors.woocommerce.addToCartText"';
+
+			if ( WC()->cart->is_empty() ) {
```
This is causing me an error when saving posts. It looks like `WC()->cart` is null, so the `is_empty()` method doesn't exist. Can you reproduce as well?

![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/bc9e1007-121e-468d-aee1-e06a0d491d2c)

gigitux committed:
improve logic

Aljullu commented on assets/js/atomic/blocks/product-elements/button/style.scss:
```diff
@@ -50,6 +61,17 @@
 		margin-left: auto;
 		white-space: normal;
 		word-break: break-word;
+		width: 150px;
```
Nit: we could add `overflow: hidden` here. Otherwise it felt to me like the text was coming from "outside" the button.

Aljullu reviewed (commented):
This is working great, really exciting to see how this is shaping up! I left some comments related to the code below, but overall it's looking good.

gigitux committed:
refactor logic

gigitux committed:
refactor code

gigitux commented:
Thanks, @Aljullu, I had a pair session with @luisherranz that explained my wrong approach. Thanks to his help, I refactored this code and implemented the animation when the quantity of the product is updated from the Mini Cart as well.

Aljullu commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -86,20 +124,108 @@ const productButtonSelectors = {
+			state,
+			context,
+		}: {
+			selecotrs: any;
```
There is a typo here.

Aljullu commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -86,20 +124,108 @@ const productButtonSelectors = {
+		hasCartLoaded: ( { state }: { state: State } ) => {
+			return state.woocommerce.cart !== undefined;
+		},
+		numberOfItems: ( {
```
Given that this is a function, I think it would be great to start its name with a verb, something like `getNumberofItems()`. What do you think?

Aljullu commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -86,20 +124,108 @@ const productButtonSelectors = {
+		} ) => {
+			const { context, selectors } = store;
+
+			const isFreshLoad =
+				selectors.woocommerce.hasCartLoaded( store ) &&
+				context.woocommerce.initialNumberOfItems !==
+					selectors.woocommerce.numberOfItems( store ) &&
+				context.woocommerce.numberOfItems !==
+					selectors.woocommerce.numberOfItems( store );
+
+			const isFromCart =
+				selectors.woocommerce.numberOfItemsPrev( store ) !==
+					undefined &&
+				selectors.woocommerce.numberOfItemsPrev( store ) !==
+					selectors.woocommerce.numberOfItems( store ) &&
+				selectors.woocommerce.numberOfItems( store ) !==
+					context.woocommerce.numberOfItems;
```
Nit: given that we are calling `numberOfItems()` several times, could we call it at the beginning of the function and save its value into a variable? Same with `numberOfItemsPrev()`. I feel it would make the code easier to read.

Aljullu commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -86,20 +124,108 @@ const productButtonSelectors = {
+
+			return product?.quantity || 0;
+		},
+		numberOfItemsPrev: ( {
```
I would change this one to start with `get` too, so it's clear it's a function.

Aljullu reviewed (commented):
As mentioned in the previous review, this is looking awesome, I'm very excited for this to be merged. :clap: 

I left some comments inline, but they are mostly about coding style, so feel free to disregard them if you don't agree.

Also, I noticed that using the `woocommerce_add_to_cart_quantity` filter doesn't work anymore. Steps to reproduce:

1. In any PHP file, add this code:

```PHP

// Function to set the quantity of items added to cart
function custom_add_to_cart_quantity( $quantity, $product_id ) {
	return 10;
}
add_filter( 'woocommerce_add_to_cart_quantity', 'custom_add_to_cart_quantity', 10, 2 );
```
2. In a page with the Products block, add a product to the cart.
3. Notice only one product is added, instead of `10` as specified in the filter from step 1.

Can you reproduce as well?

gigitux committed:
address feedback about code style

gigitux committed:
add support for woocommerce_add_to_cart_quantity

gigitux commented:
> As mentioned in the previous review, this is looking awesome, I'm very excited for this to be merged. 👏
> 
> I left some comments inline, but they are mostly about coding style, so feel free to disregard them if you don't agree.
> 
> Also, I noticed that using the `woocommerce_add_to_cart_quantity` filter doesn't work anymore. Steps to reproduce:
> 
> 1. In any PHP file, add this code:
> 
> ```
> // Function to set the quantity of items added to cart
> function custom_add_to_cart_quantity( $quantity, $product_id ) {
> 	return 10;
> }
> add_filter( 'woocommerce_add_to_cart_quantity', 'custom_add_to_cart_quantity', 10, 2 );
> ```
> 
> 2. In a page with the Products block, add a product to the cart.
> 3. Notice only one product is added, instead of `10` as specified in the filter from step 1.
> 
> Can you reproduce as well?

Nice catch! I added the support to the filter with [c6acd18](https://github.com/woocommerce/woocommerce-blocks/pull/10351/commits/c6acd183a6a48144bb9985f0b2bc144c2e1d21c4)

luisherranz committed:
Fix animation flickering

luisherranz committed:
Introduce wp-effect, reduce the amount of numberOfItem variables to 2 and consolidate animation status

luisherranz commented:
Hey Luigi, I think the number of variables storing `numberOfItems` was getting out of hand so I tried a new approach [based on this logic](https://excalidraw.com/#json=y4x3o_KTZPeeLmXb1rMjX,yEX_WySmqiWBqgn234VRgw). It took me a while to figure out, but I think it's working great. I also introduced a `wp-effect` to simplify the logic that starts the animation. I forgot about `wp-effect` the other day when we looked at it.

I made a video to explain everything:

<div>
    <a href="https://www.loom.com/share/19478a6e951442f09f5902d3fb20c68e">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/19478a6e951442f09f5902d3fb20c68e-with-play.gif">
    </a>
  </div>

https://www.loom.com/share/19478a6e951442f09f5902d3fb20c68e

I still want to take a look at how to await the population of the store in the `AddToCart` action, but let me know what you think.

gigitux committed:
add support for added class

gigitux commented:
> Hey Luigi, I think the number of variables storing `numberOfItems` was getting out of hand so I tried a new approach [based on this logic](https://excalidraw.com/#json=y4x3o_KTZPeeLmXb1rMjX,yEX_WySmqiWBqgn234VRgw). It took me a while to figure out, but I think it's working great. I also introduced a `wp-effect` to simplify the logic that starts the animation. I forgot about `wp-effect` the other day when we looked at it.
> 
> I made a video to explain everything:
> 
> [ ![](https://camo.githubusercontent.com/fa13edd60b38b893abbe74f3360203d99125b792f0084efaccceed89d7a9de1c/68747470733a2f2f63646e2e6c6f6f6d2e636f6d2f73657373696f6e732f7468756d626e61696c732f31393437386136653935313434326630396635393032643366623230633638652d776974682d706c61792e676966) ](https://www.loom.com/share/19478a6e951442f09f5902d3fb20c68e)
> https://www.loom.com/share/19478a6e951442f09f5902d3fb20c68e
> 
> I still want to take a look at how to await the population of the store in the `AddToCart` action, but let me know what you think.

Thanks for your work and your great explanation! I forget entirely about the `wp-effect` 😅

With [1333cc7](https://github.com/woocommerce/woocommerce-blocks/pull/10351/commits/1333cc70c9100c3a5a1c62682f6ce16d9d0c14cf) I re-added the `added` class support. Currently, the class `added` is only added if the user adds the product by clicking the button. I think that it makes sense replicate the same behavior, but feel free to give your feedback.

gigitux commented:
Regarding the other improvements that you mentioned:
- using the scope.
- having one source of truth, so waiting for the store to be updated in action.

Should we continue to work on it, or can we start to merge this and work on them with follow-up issues/PRs?

luisherranz committed:
Remove unnecessary selector

luisherranz committed:
Don't fetch cart if it was already fetched

luisherranz commented:
> having one source of truth, so waiting for the store to be updated in action.

It works fine 🙂

<div>
    <a href="https://www.loom.com/share/3dde7d89d1c8496fa8c6d1b4a2652f57">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/3dde7d89d1c8496fa8c6d1b4a2652f57-with-play.gif">
    </a>
  </div>

https://www.loom.com/share/3dde7d89d1c8496fa8c6d1b4a2652f57

> can we start to merge this and work on them with follow-up issues/PRs?

Yeah, let's merge this PR now. I'd like to take a look at the need for React and `NoticeBanner`, but we can do it in another PR.

gigitux committed:
remove added class

gigitux commented:
Regarding the `.added` class I searched across the extensions and it seems that any extension uses the `.added` class in their logic. I think that for now, we can remove it, and if it is necessary, we can easily re-add it.

I removed the logic for the `.added` class with 49651c23143d5345a61333a0ae9bdd702dbb9b29.

gigitux commented:
If you think that everything is ready, could you approve the PR? @luisherranz 🙇

luisherranz reviewed (approved):
Yes, let's merge this 🙂
